README.md

```diff
@@ -1,209 +1,404 @@
+  // single EchoResponse protobuf message in response.
+  rpc Get(EchoRequest) returns (EchoResponse) {}
+}
+
+// The EchoRequest protobuf message definition.
+message EchoRequest {
+  // The text of a message to be echoed.
+  string text = 1;
+}
+
+// The EchoResponse protobuf message definition.
+message EchoResponse {
+  // The text of an echo response.
+  string text = 1;
+}
+```
+
+#### Generating the Service and Client Code
+
+Once we have defined our service we can generate the necessary code to implement
+our service and client.
+
+Models (such as `EchoRequest` and `EchoResponse`) are generated using the
+`protoc` plugin provided by [SwiftProtobuf][swift-protobuf]. Assuming the
+above definition for the Echo service and models were saved as `echo.proto` and
+the `protoc-gen-swift` plugin is available in your `PATH` then the following
+will generate the models and write them to `echo.pb.swift` in the current
+directory:
+
+```sh
+protoc echo.proto --swift_out=.
+```
+
+Swift gRPC, provides a plugin (`protoc-gen-swiftgrpc`) to generate the client
+and server for the `Echo` service defined above. It can be invoked to produce
+`echo.grpc.swift` as such:
+
+```sh
+protoc echo.proto --swiftgrpc_out=.
+```
+
+By default both the client and service code is generated (see [Plugin
+Options](#plugin-options) for more details).
+
+#### Implementing the Service Code
+
+The generated service code includes a protocol called `Echo_EchoProvider` which
+defines the set of RPCs that the service offers. The service can be implemented
+by creating a class which conforms to the protocol.
+
+```swift
+import GRPC
+import NIO
+
+class EchoProvider: Echo_EchoProvider {
+  func get(request: Echo_EchoRequest, context: StatusOnlyCallContext) -> EventLoopFuture<Echo_EchoResponse> {
+    let response = Echo_EchoResponse.with {
+      $0.text = "Swift echo get: \(request.text)"
+    }
+    return context.eventLoop.makeSucceededFuture(response)
+  }
+}
+```
+
+#### Using the Service Provider and Generated Client
+
+Now that we have implemented our service code and generated a client, we can put
+it all together.
+
+First we need the appropriate imports:
+
+```swift
+import GRPC
+import NIO
+import Foundation
+```
+
+Since this is just a locally hosted example we'll run the server on localhost
+port 8080.
+
+```
+let host = "localhost"
+let port = 8080
+```
+
+First we need to start the server and provide the Echo service using the
+`EchoProvider` we implemented above. We create a [multhithreaded event loop
+group][nio-ref-elg] which will spawn a single thread to run events on the
+server. Note that you can also use `System.coreCount` to get the number of
+logical cores on your system.
+
+We also `wait` for the server to start before setting up a client.
+
+```swift
+let server = try Server.start(
+  hostname: host,
+  port: port,
+  eventLoopGroup: MultiThreadedEventLoopGroup(numberOfThreads: 1),
```

Comment:
I assume this event loop group would never be shut down? (Fine for an example and most servers that run indefinitely, I guess)

Reply:
That's correct and this should definitely be fixed! 

In parallel, I've been doing the work to get NIOTS integrated. This includes giving the user an option to specify which model to use (NIO vs NIOTS), which is actually "always NIO" or "NIOTS when possible, otherwise NIO". Since that option fits quite nicely into `Configuration` (this needs to be done for the `Server` too) and the `EventLoopGroup` depends on that option, it might make more sense for the client/server to own the ELG, and shut it down when it closes.

What do you think?

Reply:
Alternatively, we could add an argument like this to `Configuration`:

```swift
enum OperationMode {
  case nio(eventLoopGroup: EventLoopGroup)
  case nioTS
}
```

Not sure if that would work, though. I'm not feeling too strongly, as long as we can have compatibility with Vapor (although that would probably focused on Linux and thus regular NIO, anyway). I guess if NIOTS makes it even easier for client users because they don't have to worry about the event loop group, that would be great.

Reply:
What are the requirements for compatibility with Vapor? 

To clarify: NIOTS needs an ELG but it's a `NIOTSEventLoopGroup` which is incompatible with `{Client,Server}Bootstrap`.

It will be easier to discuss this in the context of the NIOTS PR. We can then circle back to this example/README.

Reply:
Good question. I guess Vapor users would need to use the non-TS version for now, anyway. We can defer this conversation in any case.